barretenberg/cpp/src/barretenberg/common/thread.cpp

```diff
@@ -127,3 +127,76 @@ void run_loop_in_parallel(size_t num_points,
         func(start, end);
     });
 };
+
+/**
+ * @brief Split a loop into several loops running in parallel based on operations in 1 iteration
+ *
+ * @details Splits the num_points into appropriate number of chunks to do parallel processing on and calls the function
+ * that should contain the work loop, but only if it's worth it
+ * @param num_points Total number of elements
+ * @param func A function or lambda expression with a for loop inside, for example:
+ * [](size_t start, size_t end){for (size_t i=start; i<end; i++){(void)i;}}
+ * @param finite_field_additions_per_iteration
+ * @param finite_field_multiplications_per_iteration
+ * @param finite_field_inversions_per_iteration
+ * @param group_element_additions_per_iteration Projective addition number
+ * @param group_element_doublings_per_iteration Projective doubling number
+ * @param scalar_multiplications_per_iteration
+ * @param sequential_copy_ops_per_iteration Field element (16 byte) sequential copy number
+ */
+void run_loop_in_parallel_if_effective(size_t num_points,
+                                       const std::function<void(size_t, size_t)>& func,
+                                       size_t finite_field_additions_per_iteration,
+                                       size_t finite_field_multiplications_per_iteration,
+                                       size_t finite_field_inversions_per_iteration,
+                                       size_t group_element_additions_per_iteration,
+                                       size_t group_element_doublings_per_iteration,
+                                       size_t scalar_multiplications_per_iteration,
+                                       size_t sequential_copy_ops_per_iteration)
+{
+    // Rough cost of operations:
+    constexpr size_t FF_ADDITION_COST = 4;
+    constexpr size_t FF_MULTIPLICATION_COST = 21;
+    constexpr size_t FF_INVERSION_COST = 7000;
+    constexpr size_t GE_ADDITION_COST = 350;
+    constexpr size_t GE_DOUBLING_COST = 194;
+    constexpr size_t SM_COST = 50000;
+    constexpr size_t SEQ_COPY_COST = 3;
+    constexpr size_t PARALLEL_FOR_COST = 376000;
```

Comment:
Does `PARALLEL_FOR_COST` not depend on the number of cpus?

Reply:
A good question, let me check